lib/tree.rb

```diff
@@ -2,61 +2,160 @@ class TreeNode
+  # Space Complexity: same as time complexity for stack calls
+  def add(key, value = nil)
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, key, value)
+    if current.nil?
+      current = TreeNode.new(key, value)
+    elsif key <= current.key
+      current.left = add_helper(current.left, key, value)
+    else
+      current.right = add_helper(current.right, key, value)
+    end
+
+    return current
+  end
+
+  # Time Complexity: O(log(n))/O(n)
+  # Space Complexity: same as time complexity
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
+  end
+
+  def find_helper(current, key)
+    return nil unless current
+    if key == current.key
+      return current.value
+    elsif key < current.key
+      find_helper(current.left, key)
+    else
+      find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity:O(n) we traverse all the nodes once
+  # Space Complexity:O(n) for stack calls + array space(O(2n)) 
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node, array)
+    return array unless node
+
+    inorder_helper(node.left, array)
+    array << { :key => node.key, :value => node.value }
+    inorder_helper(node.right, array)
+  end
+
+  # Time Complexity:O(n) 
+  # Space Complexity:O(n) 
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, array)
+    return array unless node
+
+    array << { :key => node.key, :value => node.value }
+    preorder_helper(node.left, array)
+    preorder_helper(node.right, array)
+  end
+
+  # Time Complexity:O(n) 
+  # Space Complexity:O(n) 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  def postorder_helper(node, array)
+    return array unless node
+
+    postorder_helper(node.left, array)
+    postorder_helper(node.right, array)
+    array << { :key => node.key, :value => node.value }
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) Have to traverse all the nodes once to find the one with the max height
+  # Space Complexity:O(n) for stack calls
   def height
-    raise NotImplementedError
+    height_helper(@root)
+  end
+
+  def height_helper(current)
+    return 0 unless current
+
+    left_ht = height_helper(current.left) + 1
+    right_ht = height_helper(current.right) + 1
+    return [left_ht, right_ht].max
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) traverse all the nodes once
+  # Space Complexity: O(n^2) for array, queue and call stacks??? This can be done iteratively too with while loop. 
   def bfs
```

Comment:
👍 You can fix the time complexity by using Ruby's built-in Queue class.  That would bring it down to O(n)